### PR TITLE
ALIGNN test on WBM data

### PR DESCRIPTION
**Pull request by pbenner**
Added ALIGNN test on WBM data, achieving a MAE of 0.0916. The code needs some double checking to make sure that  matbench-discovery is correctly used. coGN would be the next step.

### CODE REVIEW

**Comment by janosh** [comment]
This is great, thank you @pbenner! 🎉 

Sorry for the slow reply here. Was about to hit send this morning when the day suddenly turned super packed. Needless to say, I'm excited about this 1st external model submission to Matbench Discovery!

I'd thought about running ALIGNN myself but expected it to be similar to MEGNet and so in the end decided against it. But looks like I was wrong. Performance is quite a bit better than expected!

I'll refactor the code here a bit and then start updating all the site metrics. You didn't by any chance log the training and test runs for ALIGNN with `wandb`?

**Comment by pbenner** [comment]
Sorry no, had some minor trouble with wandb and excluded it, but should be easy to add to the existing code.

**Comment by janosh** [comment]
@pbenner Just out of curiosity, did you try the pre-trained `mp_e_form_alignnn`?

```py
from alignn.pretrained import get_figshare_model

model = get_figshare_model("mp_e_form_alignnn")
```

If not, I'll run it. Would be good to know how it compares to the ALIGNN specifically trained for this benchmark.

Also, looks like we can avoid writing all structures to disk as POSCAR and then reading them in again as Jarvis Atoms. We can convert them directly using `JarvisAtomsAdaptor` (see [fffd6d7](https://github.com/janosh/matbench-discovery/pull/37/commits/fffd6d73b4134833786c1a53a5fbf08a73aed0cd)).

```py
from pymatgen.io.jarvis import JarvisAtomsAdaptor
```

**Comment by janosh** [comment]
@pbenner Could you briefly explain the need for `alignn-2023.01.10.patch` (formerly `train_alignn_patch.txt`)?

**Comment by pbenner** [comment]
I didn't know about the *JarvisAtomsAdaptor*, that's a good hint!

I tried the pretrained model and got a MAE of 0.1095, so slightly worse. However, I'm not sure if I have to apply some corrections to the energies, so training the model myself seemed safer. In general, predictive performance is quite sensitive to the number of training samples, so it seems best to me to always use the exact same training data instead of pretrained models (at least if the goal is to compare model architectures).

The ALIGNN patch has multiple purposes. First, it fixes some issues when there is no test set [(see ALIGNN Issue #104)](https://github.com/usnistgov/alignn/issues/104#issue-1723978225). In this case, we actually forked a test set, but it might be better to use the entire data, as mentioned above, especially if the test set by chance contains some important outliers. Second, the *Checkpoint()* handler in ALIGNN does not define a score name [(see ALIGNN train.py)](https://github.com/usnistgov/alignn/blob/46334500cac9833125b3e444d65d0246e692bd61/alignn/train.py#L851), so it will just save the last two models during training. With this patch, also the best model in terms of accuracy on the validation set is saved, which is the one I use for predictions. This is important, because I used a relatively large *n_early_stopping* in case the validation accuracy shows a double descent [(see Figure 10)](https://arxiv.org/pdf/1912.02292.pdf). Maybe one has to consider here also how individual models are trained, since longer training might be beneficial.

**Comment by janosh** [comment]
I also ran the pre-trained ALIGNN before merging this PR just to check how it compares. The MAE doesn't tell the whole story. As a classifier, it significantly underperforms the ALIGNN trained from scratch for MBD.

#### Raw pre-trained ALIGNN (`mp_e_form_alignn`)

photo_2023-06-05 17 49 37

#### After subtracting the legacy corrections ALIGNN was trained on and adding the new MP 2020 corrections:

Slightly better classification performance but still bad.

photo_2023-06-05 17 49 33

---

Btw, you can find both the old and new MP corrections in `matbench_discovery.data.df_wbm` so updating model preds is just two lines: subtract `e_correction_per_atom_mp_legacy` and add `e_correction_per_atom_mp2020`.

Speaking of checkpoints, could you send me the checkpoint that was used to make the predictions in this PR? Would be good to provide that for reproducibility.

**Comment by pbenner** [comment]
Sure, the model is relatively small, so I just added it to the repo: [Alignn best model](https://github.com/pbenner/matbench-discovery/raw/main/models/alignn/data_train_result/model_best.pth)

**Comment by CompRhys** [comment]
@pbenner do you know whether it's possible to run your alignn model in a uip-gnn setup where it first relaxes itself according to it's own predictions? In some of the metrics we see what we've been calling a characteristic fire-hook curve which we believe is a function of the task mismatch seen in the IS2RE task that is mostly resolved with uip-relaxed-IS2RE.

**Comment by janosh** [comment]
@CompRhys The original ALIGNN doesn't predict forces. They released a [force field version of ALIGNN a year later in 2022](https://arxiv.org/abs/2209.05554) which looks similar to M3GNet. You could use the original ALIGNN to do repeated energy preds to get finite-difference forces but every time I've seen that it worked poorly. The ALIGNN FF would be interesting to add though.

**Comment by pbenner** [comment]
Just a small update. I am currently testing the ALIGNN-FF model for computing relaxed structures:
https://github.com/pbenner/matbench-discovery/blob/main/models/alignn/alignn_relax.py
It's quite slow, so I have to see how to get results within some reasonable time frame.

**Comment by janosh** [comment]
Nice! 🎉 

For CHGnet and M3GNet, I used `np.array_split()` to split the unrelaxed structures into 100 subsets and then launched slurm jobs to relax each subset in parallel. That way it takes ~2 hours to get a complete set of predictions.

https://github.com/janosh/matbench-discovery/blob/4a90dee85965aebbcfcd547353c548eeb19651ce/models/chgnet/test_chgnet.py#L67-L69

**Comment by pbenner** [comment]
Thanks! It takes around 3 days on 4 A100 GPUs. Results should be done next week.

**Comment by janosh** [comment]
Sweet, looking forward to it!

**Comment by janosh** [comment]
> The relaxed predictions are finally done! You can find the results here:

Excellent. That's great news!

> I'm not filing a pull request, because my repository contains the full set of relaxed structures, which you might want to upload to figshare.

Sounds good. I'll start prepping a PR with your code and results then?

> the ones that converge poorly take several days to finish.

What?! That sounds problematic. You're saying some of the 100 test set batches (~2500 structures) took several days? If so, suggests lack of robustness to data variety in the model.
